enhancements/ingress/host-network-configurable-ports.md

```diff
@@ -0,0 +1,298 @@
+---
+title: ingress-host-network-configurable-ports
+
+  authors:
+- "@elbehery"
+- "@sherine-k"
+ 
+reviewers:
+- "@alebedev87"
+- "@frobware"
+- "@Miciah"
+
+approvers:
+- "@knobunc"
+- "@tjungblu"
+
+
+  creation-date: 2022-01-10
+  
+  last-updated: 2022-01-10
+
+  status: implementable
+
+  see-also:
+
+  replaces:
+
+  superseded-by:
+
+---
+
+# Ingress Host Network Configurable Ports
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement updates the Ingress Controller API for the HostNetwork strategy, specifying the ports to be used by the Ingress Controller pods. This allows running several instances of the ingress controller on the same node using the HostNetwork strategy, therefore avoiding port binding conflicts.
+
+## Motivation
+
+While utilising the HostNetwork strategy for ingress controllers, the ports 80, 443, 10443, 10444, 1936 on the host are bound by HAProxy for http, https, no_sni, sni and stats respectively. 
+
+These ports cannot be used by more than one process. Therefore, when more than one Ingress Controller with HostNetwork strategy is deployed on a cluster, each Ingress Controller needs to deploy its pods to cluster nodes that are not already used by another Ingress Controller.
+
+This results sometimes in increasing the number of cluster nodes, not to meet performance requirements, but for the sole purpose of having nodes with http, https and stats ports that are available to use for a new Ingress Controller that the cluster administrators are introducing.
+
+### Goals
+
+Enable running multiple sets of ingress controllers on the same node with `HostNetwork` endpoint strategy. The approach is to provide, in the ingress controller's spec, the port numbers to be used by its pods for:
+- HTTP
+- HTTPS
+- Stats
+
+### Non-Goals
+
+- Allowing configuring custom ports on the ingress controllers that utilise any endpoint publishing strategy other than `HostNetwork`.
+  
+
+- Allowing configuring custom ports on the ingress controllers for `SNI` and `NO_SNI` ports.
+
+
+## Proposal
+
+Configuring http, https and stats port for the ingress controllers that use the `HostNetwork` endpoint publishing strategy.
+
+
+Update the IngressController API by adding an optional `HostNetworkParameters` field with type `*HostNetworkParameters` to the `HostNetworkStrategy` struct:
+
+```go
```

Comment:
This multiple level of structs for the configuration seems redundant since the structs contain a single field. The _HTTP_, _HTTPS_ and _STATS_ port configuration can be moved into `HostNetworkStrategy` directly.

Reply:
I agree with the idea of the simplification. But I think it can be done in many ways.

1) What Arjun proposed:
```yaml
endpointPublishingStrategy:
  type: HostNetwork
  hostNetwork:
    httpPort: 10080
    httpsPort: 10443
    statsPort: 11936
```
2) Similar to Arjun's one but with 1 level more to avoid `Port` suffix:
```yaml
endpointPublishingStrategy:
  type: HostNetwork
  hostNetwork:
    ports:
      http: 10080
      https: 10443
      stats: 11936
```
3) What Sherine put into EP:
```yaml
endpointPublishingStrategy:
  type: HostNetwork
  hostNetwork:
     hostNetworkParameters:
       httpPort: 10080
       httpsPort: 10443
       statsPort: 11936
```

Reply:
Removing the `hostNetworkParameters` level makes sense too.
I vote for # 1